core/impl/src/main/java/com/blazebit/persistence/impl/AbstractCTECriteriaBuilder.java

```diff
@@ -201,38 +209,83 @@ public CTEInfo createCTEInfo() {
     }
 
     protected List<String> prepareAndGetAttributes() {
-        List<String> attributes = new ArrayList<String>(bindingMap.size());
-        for (Map.Entry<String, Integer> bindingEntry : bindingMap.entrySet()) {
-            final String attributeName = bindingEntry.getKey();
-
-            ExtendedAttribute attributeEntry = attributeEntries.get(attributeName);
-            List<Attribute<?, ?>> attributePath = attributeEntry.getAttributePath();
-            attributes.add(attributeName);
-
-            if (JpaMetamodelUtils.isJoinable(attributePath.get(attributePath.size() - 1))) {
-                // We have to map *-to-one relationships to their ids
-                EntityType<?> type = mainQuery.metamodel.entity(attributeEntry.getElementClass());
-                Attribute<?, ?> idAttribute = JpaMetamodelUtils.getSingleIdAttribute(type);
+        final Queue<String> attributeQueue = new ArrayDeque<>(bindingMap.keySet());
+        while (!attributeQueue.isEmpty()) {
+            final String attributeName = attributeQueue.remove();
+            Integer tupleIndex = bindingMap.get(attributeName);
+
+            final ExtendedAttribute attributeEntry = attributeEntries.get(attributeName);
+            final List<Attribute<?, ?>> attributePath = attributeEntry.getAttributePath();
+            final JpaMetamodelAccessor jpaMetamodelAccessor = mainQuery.jpaProvider.getJpaMetamodelAccessor();
+            final Attribute<?, ?> lastAttribute = attributePath.get(attributePath.size() - 1);
+
+            if (jpaMetamodelAccessor.isJoinable(lastAttribute) || lastAttribute.getPersistentAttributeType() == Attribute.PersistentAttributeType.EMBEDDED) {
```

Comment:
Perhaps I should use `isComponentType` here? Any suggestions as to how to improve this check across ORMs?

Reply:
Seems fine to me